### PR TITLE
Add support for virtual environments in addition to conda environments

### PR DESCRIPTION
**Pull request by zmbc**
Tested this with all combinations of pandas/local Dask/local Spark/distributed Dask/distributed Spark.

### CODE REVIEW

File: src/person_linkage_case_study_utils/snakemake_utils.py [hunk 1]
```diff
@@ -127,4 +127,6 @@ def get_directory_wrapper_if_necessary(papermill_params):
 # because it does not generate incremental output, nor output notebooks when there is
 # an error. See https://github.com/snakemake/snakemake/pull/2857
 def dict_to_papermill(d):
-    return " ".join([f"-p {k} {v}" for k, v in d.items()])
+    from shlex import quote
```
Comment (patricktnast) [inline src/person_linkage_case_study_utils/snakemake_utils.py]: nit--do you have some reason for not wanting to put this import at the top?

File: src/person_linkage_case_study_utils/start_spark_workers.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+parser.add_argument("conda_prefix")
+parser.add_argument("venv_path")
+parser.add_argument("master_url")
+parser.add_argument("local_dir")
+
+args = parser.parse_args()
+
+host = socket.getfqdn()
+
+if args.conda_path != "":
```
Comment (patricktnast) [inline src/person_linkage_case_study_utils/start_spark_workers.py]: could you have the same `"None"` situation as you check for in `spark_master.py`?

File: src/person_linkage_case_study_utils/distributed_compute.py [hunk 3]
```diff
@@ -810,8 +810,12 @@ def to_pyarrow_large_string(df):
 
 def start_spark_cluster(
     local: bool,
+    singularity_path: str | pathlib.Path,
```
Comment (patricktnast) [inline src/person_linkage_case_study_utils/distributed_compute.py]: nit: do you use anything else from pathlib? i.e. why not import as `from pathlib import Path`

**Review by patricktnast — APPROVED** [review]
looks good AFAICT

File: src/person_linkage_case_study_utils/snakemake_utils.py (hunk 1 above)
Comment (zmbc) [reply src/person_linkage_case_study_utils/snakemake_utils.py]: Nope!

File: src/person_linkage_case_study_utils/start_spark_workers.py (hunk 2 above)
Comment (zmbc) [reply src/person_linkage_case_study_utils/start_spark_workers.py]: Yes! Good catch!

**Pull request merged**